Unable to run unit tests that reference MsalService

### Core Library

MSAL.js (@azure/msal-browser)

### Core Library Version

3.3.0

### Wrapper Library

MSAL Angular (@azure/msal-angular)

### Wrapper Library Version

3.0.6

### Public or Confidential Client?

Public

### Description

Unable to run unit tests due to a missing file in the release. Downgraded to 3,2,0 (and @azure/msal-angular@3.0.5) and the unit test successfully runs.

jest@28.1.3
jest-preset-angular@12.2.3

Angular CLI: 15.2.8
Node: 18.12.1
Package Manager: npm 8.19.2
OS: win32 x64

Angular: 15.2.9
... animations, common, compiler, compiler-cli, core, forms
... localize, platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1502.8
@angular-devkit/build-angular   15.2.8
@angular-devkit/core            15.2.8
@angular-devkit/schematics      15.2.8
@angular/cli                    15.2.8
@schematics/angular             15.2.8
rxjs                            7.8.1
typescript                      4.8.4


### Error Message

  ● Test suite failed to run

    Cannot find module './index-9800158f.js' from 'node_modules/@azure/msal-browser/lib/msal-browser.cjs'

    Require stack:
      node_modules/@azure/msal-browser/lib/msal-browser.cjs
      node_modules/@azure/msal-angular/fesm2020/azure-msal-angular.mjs
      src/app/auth/auth.state.spec.ts

### Msal Logs

Not applicable

### MSAL Configuration

```javascript
Not Applicable
```


### Relevant Code Snippets

```javascript
See stack trace
```


### Reproduction Steps

`$ jest`

### Expected Behavior

Jest runs

### Identity Provider

Azure AD / MSA

### Browsers Affected (Select all that apply)

None (Server), Other

### Regression

@azure/msal-browser 3.2.0

### Source

Internal (Microsoft)

I also have this issue (when running my unit tests with Jest) ...

```
Cannot find module './index-9800158f.js' from '../../node_modules/@azure/msal-browser/lib/msal-browser.cjs' 
```

@kai-dorschner-twinsity posted already the deep dive into code I also did. I wonder how this could happen where that mentioned `index-9800158f.js` should be. 

@kebryansg Can you please post complete example of your `jest.config.js`?

I tried out your proposed workaround. But, it doesn't work for me. Maybe I did it wrong... Or it's different for me anyway. I have a mono repo setup with NX and a typescript version of jest config => `jest.config.ts`.

My **(not working)** version (should be fine from my point of view considering also official Jest documentation about `moduleNameMapper`):

```
const { getJestProjects } = require('@nx/jest');

export default {
    moduleNameMapper: {
        '@azure/msal-browser': '<rootDir>/node_modules/@azure/msal-browser/lib/msal-browser.js',
    },
    projects: getJestProjects()
};
```


https://jestjs.io/docs/configuration#modulenamemapper-objectstring-string--arraystring

--- a/change/@azure-msal-browser-be125892-2c84-4c9d-89d8-bb5c7f5490c9.json
+++ b/change/@azure-msal-browser-be125892-2c84-4c9d-89d8-bb5c7f5490c9.json
@@ -1,0 +1,7 @@
+{
+  "type": "patch",
+  "comment": "Add \"inlineDynamicImports\" to rollup config to fix CJS build. Make factory methods basic functions #6615",
+  "packageName": "@azure/msal-browser",
+  "email": "kshabelko@microsoft.com",
+  "dependentChangeType": "patch"
+}

--- a/lib/msal-browser/rollup.config.js
+++ b/lib/msal-browser/rollup.config.js
@@ -8,7 +8,9 @@ import typescript from "@rollup/plugin-typescript";
 import terser from "@rollup/plugin-terser";
 import pkg from "./package.json";
 
-const libraryHeader = `/*! ${pkg.name} v${pkg.version} ${new Date().toISOString().split("T")[0]} */`;
+const libraryHeader = `/*! ${pkg.name} v${pkg.version} ${
+    new Date().toISOString().split("T")[0]
+} */`;
 const useStrictHeader = "'use strict';";
 const fileHeader = `${libraryHeader}\n${useStrictHeader}`;
 
@@ -27,18 +29,18 @@ export default [
         },
         treeshake: {
             moduleSideEffects: false,
-            propertyReadSideEffects: false
+            propertyReadSideEffects: false,
         },
         external: [
             ...Object.keys(pkg.dependencies || {}),
-            ...Object.keys(pkg.peerDependencies || {})
+            ...Object.keys(pkg.peerDependencies || {}),
         ],
         plugins: [
             typescript({
                 typescript: require("typescript"),
-                tsconfig: "tsconfig.build.json"
-            })
-        ]
+                tsconfig: "tsconfig.build.json",
+            }),
+        ],
     },
     {
         input: "src/index.ts",
@@ -49,21 +51,22 @@ export default [
                 format: "cjs",
                 banner: fileHeader,
                 sourcemap: true,
-                entryFileNames: "msal-browser.cjs"
+                entryFileNames: "msal-browser.cjs",
+                inlineDynamicImports: true,
             },
         ],
         plugins: [
             nodeResolve({
                 browser: true,
-                resolveOnly: ["@azure/msal-common", "tslib"]
+                resolveOnly: ["@azure/msal-common", "tslib"],
             }),
             typescript({
                 typescript: require("typescript"),
                 tsconfig: "tsconfig.build.json",
                 sourceMap: true,
-                compilerOptions: { outDir: "./lib" }
-            })
-        ]
+                compilerOptions: { outDir: "./lib" },
+            }),
+        ],
     },
     {
         input: "src/index.ts",
@@ -76,21 +79,21 @@ export default [
                 banner: fileHeader,
                 inlineDynamicImports: true,
                 sourcemap: true,
-                entryFileNames: "msal-browser.js"
-            }
+                entryFileNames: "msal-browser.js",
+            },
         ],
         plugins: [
             nodeResolve({
                 browser: true,
-                resolveOnly: ["@azure/msal-common", "tslib"]
+                resolveOnly: ["@azure/msal-common", "tslib"],
             }),
             typescript({
                 typescript: require("typescript"),
                 tsconfig: "tsconfig.build.json",
                 sourceMap: true,
-                compilerOptions: { outDir: "./lib" }
-            })
-        ]
+                compilerOptions: { outDir: "./lib" },
+            }),
+        ],
     },
     {
         // Minified version of msal
@@ -104,24 +107,24 @@ export default [
                 banner: useStrictHeader,
                 inlineDynamicImports: true,
                 sourcemap: false,
-            }
+            },
         ],
         plugins: [
             nodeResolve({
                 browser: true,
-                resolveOnly: ["@azure/msal-common", "tslib"]
+                resolveOnly: ["@azure/msal-common", "tslib"],
             }),
             typescript({
                 typescript: require("typescript"),
                 tsconfig: "tsconfig.build.json",
                 sourceMap: false,
-                compilerOptions: { outDir: "./lib" }
+                compilerOptions: { outDir: "./lib" },
             }),
             terser({
                 output: {
-                    preamble: libraryHeader
-                }
-            })
-        ]
-    }
+                    preamble: libraryHeader,
+                },
+            }),
+        ],
+    },
 ];

--- a/lib/msal-browser/src/app/PublicClientApplication.ts
+++ b/lib/msal-browser/src/app/PublicClientApplication.ts
@@ -20,7 +20,7 @@ import {
 } from "@azure/msal-common";
 import { EndSessionRequest } from "../request/EndSessionRequest";
 import { SsoSilentRequest } from "../request/SsoSilentRequest";
-import { ControllerFactory } from "../controllers/ControllerFactory";
+import * as ControllerFactory from "../controllers/ControllerFactory";
 import { StandardController } from "../controllers/StandardController";
 import { BrowserConfiguration, Configuration } from "../config/Configuration";
 import { StandardOperatingContext } from "../operatingcontext/StandardOperatingContext";
@@ -38,8 +38,9 @@ export class PublicClientApplication implements IPublicClientApplication {
     public static async createPublicClientApplication(
         configuration: Configuration
     ): Promise<IPublicClientApplication> {
-        const factory = new ControllerFactory(configuration);
-        const controller = await factory.createV3Controller();
+        const controller = await ControllerFactory.createV3Controller(
+            configuration
+        );
         const pca = new PublicClientApplication(configuration, controller);
 
         return pca;

--- a/lib/msal-browser/src/app/PublicClientNext.ts
+++ b/lib/msal-browser/src/app/PublicClientNext.ts
@@ -20,7 +20,7 @@ import {
 } from "@azure/msal-common";
 import { EndSessionRequest } from "../request/EndSessionRequest";
 import { SsoSilentRequest } from "../request/SsoSilentRequest";
-import { ControllerFactory } from "../controllers/ControllerFactory";
+import * as ControllerFactory from "../controllers/ControllerFactory";
 import { BrowserConfiguration, Configuration } from "../config/Configuration";
 import { EventCallbackFunction } from "../event/EventMessage";
 import { ClearCacheRequest } from "../request/ClearCacheRequest";
@@ -47,8 +47,9 @@ export class PublicClientNext implements IPublicClientApplication {
     public static async createPublicClientApplication(
         configuration: Configuration
     ): Promise<IPublicClientApplication> {
-        const factory = new ControllerFactory(configuration);
-        const controller = await factory.createController();
+        const controller = await ControllerFactory.createController(
+            configuration
+        );
         let pca;
         if (controller !== null) {
             pca = new PublicClientNext(configuration, controller);
@@ -100,8 +101,9 @@ export class PublicClientNext implements IPublicClientApplication {
      */
     async initialize(): Promise<void> {
         if (this.controller instanceof UnknownOperatingContextController) {
-            const factory = new ControllerFactory(this.configuration);
-            const result = await factory.createController();
+            const result = await ControllerFactory.createController(
+                this.configuration
+            );
             if (result !== null) {
                 this.controller = result;
             }

--- a/lib/msal-browser/src/controllers/ControllerFactory.ts
+++ b/lib/msal-browser/src/controllers/ControllerFactory.ts
@@ -6,61 +6,42 @@
 import { TeamsAppOperatingContext } from "../operatingcontext/TeamsAppOperatingContext";
 import { StandardOperatingContext } from "../operatingcontext/StandardOperatingContext";
 import { IController } from "./IController";
-import { Logger } from "@azure/msal-common";
 import { Configuration } from "../config/Configuration";
-import { version, name } from "../packageMetadata";
 
-export class ControllerFactory {
-    protected config: Configuration;
-    protected logger: Logger;
+export async function createV3Controller(
+    config: Configuration
+): Promise<IController> {
+    const standard = new StandardOperatingContext(config);
 
-    constructor(config: Configuration) {
-        this.config = config;
-        const loggerOptions = {
-            loggerCallback: undefined,
-            piiLoggingEnabled: false,
-            logLevel: undefined,
-            correlationId: undefined,
-        };
-        this.logger = new Logger(loggerOptions, name, version);
-    }
+    await standard.initialize();
 
-    async createV3Controller(): Promise<IController> {
-        const standard = new StandardOperatingContext(this.config);
+    const controller = await import("./StandardController");
+    return await controller.StandardController.createController(standard);
+}
 
-        await standard.initialize();
+export async function createController(
+    config: Configuration
+): Promise<IController | null> {
+    const standard = new StandardOperatingContext(config);
+    const teamsApp = new TeamsAppOperatingContext(config);
 
+    const operatingContexts = [standard.initialize(), teamsApp.initialize()];
+
+    await Promise.all(operatingContexts);
+
+    if (
+        teamsApp.isAvailable() &&
+        teamsApp.getConfig().auth.supportsNestedAppAuth
+    ) {
+        const controller = await import("./NestedAppAuthController");
+        return await controller.NestedAppAuthController.createController(
+            teamsApp
+        );
+    } else if (standard.isAvailable()) {
         const controller = await import("./StandardController");
         return await controller.StandardController.createController(standard);
-    }
-
-    async createController(): Promise<IController | null> {
-        const standard = new StandardOperatingContext(this.config);
-        const teamsApp = new TeamsAppOperatingContext(this.config);
-
-        const operatingContexts = [
-            standard.initialize(),
-            teamsApp.initialize(),
-        ];
-
-        await Promise.all(operatingContexts);
-
-        if (
-            teamsApp.isAvailable() &&
-            teamsApp.getConfig().auth.supportsNestedAppAuth
-        ) {
-            const controller = await import("./NestedAppAuthController");
-            return await controller.NestedAppAuthController.createController(
-                teamsApp
-            );
-        } else if (standard.isAvailable()) {
-            const controller = await import("./StandardController");
-            return await controller.StandardController.createController(
-                standard
-            );
-        } else {
-            // Since neither of the actual operating contexts are available keep the UnknownOperatingContextController
-            return null;
-        }
+    } else {
+        // Since neither of the actual operating contexts are available keep the UnknownOperatingContextController
+        return null;
     }
 }

